File: Find_Median_from_Data_Stream.java
/**
 * Problem Link: https://leetcode.com/problems/find-median-from-data-stream/
 */
class MedianFinder {
    private List<Integer> sortedArray = new ArrayList<Integer>();

    // Adds a number into the data structure.
    public void addNum(int num) {
        int left = 0;
        int right = sortedArray.size();
        while(left < right) {
            int mid = (left + right) / 2;
            if(sortedArray.get(mid) <= num) left = mid + 1;
            else right = mid;
        }
        // if(sortedArray.get(left) > num) sortedArray.add(0, num);
        // else 
        sortedArray.add(left, num);
    }

    // Returns the median of current data stream
    public double findMedian() {
        if(sortedArray.size() == 0) return 0.0;
        int mid = sortedArray.size() / 2;
        if(sortedArray.size() % 2 == 1) {
            return sortedArray.get(mid);
        } else {
            return .5 * (sortedArray.get(mid) + sortedArray.get(mid - 1));
        }
    }
};

// Your MedianFinder object will be instantiated and called as such:
// MedianFinder mf = new MedianFinder();
// mf.addNum(1);
// mf.findMedian();

File: Swift/find-median-from-data-stream.swift
/**
 * https://leetcode.com/problems/find-median-from-data-stream/
 * 
 * 
 */ 
// Date: Sun Jul 11 08:41:21 PDT 2021

class MedianFinder {

    var list: [Int]
    /** initialize your data structure here. */
    init() {
        list = []
    }
    
    func addNum(_ num: Int) {
        var left = 0
        var right = list.count
        while left < right {
            let mid = left + (right - left) / 2
            if list[mid] < num {
                left = mid + 1
            } else {
                right = mid
            }
        }
        list.insert(num, at: left)
    }
    
    func findMedian() -> Double {
        let n = self.list.count
        if n % 2 == 0 {
            return (Double(self.list[n / 2]) + Double(self.list[n / 2 - 1])) / 2
        }
        return Double(self.list[n / 2])
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * let obj = MedianFinder()
 * obj.addNum(num)
 * let ret_2: Double = obj.findMedian()
 */